packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,170 @@
+ *
+ * Note: We use `Omit` instead of `Pick` to ensure that all new options are included in this type by default. If any new properties
+ * are added to IContainerRuntimeOptionsInternal, they will be included in this type unless explicitly omitted. This will prevent
+ * us from forgetting to account for any new properties in the future.
+ */
+export type IContainerRuntimeOptionsVersionDependent = Required<
+	Omit<
+		IContainerRuntimeOptionsInternal,
+		| "chunkSizeInBytes"
+		| "maxBatchSizeInBytes"
+		| "loadSequenceNumberVerification"
+		| "summaryOptions"
+		| "compatibilityMode"
+	>
+>;
+
+/**
+ * Mapping of version-dependent options to their compatibility related configs.
+ *
+ * Each key in this map corresponds to a property in IContainerRuntimeOptionsVersionDependent.
+ * The value is an object that maps a version string to the default value for that property when using compatibilityMode of at least that version.
+ */
+const versionDependentOptionConfigMap: {
+	[K in keyof IContainerRuntimeOptionsVersionDependent]: {
+		[version: string]: IContainerRuntimeOptionsVersionDependent[K];
+	};
+} = {
+	enableGroupedBatching: {
+		"1.0.0": false,
+		"2.0.0": true,
+	},
+	compressionOptions: {
+		"1.0.0": disabledCompressionConfig,
+		"2.0.0": enabledCompressionConfig,
+	},
+	enableRuntimeIdCompressor: {
+		// For IdCompressorMode, `undefined` represents a logical state (off). However, to satisfy the Required<>
+		// constraint we need to have it defined, so we trick the type checker here.
+		"1.0.0": undefined as unknown as "on" | "delayed",
+		// We do not yet want to enable idCompressor by default since it will increase bundle sizes,
+		// and not all customers will benefit from it. Therefore, we will require customers to explicitly
+		// enable it. We are keeping it as a version-dependent option as this may change in the future.
+	},
+	explicitSchemaControl: {
+		// This option is unique since it was actually introduced before 2.0.0, but its purpose is to prevent 1.x clients from
+		// joining a session. Therefore, we will have it be `true` when the compatibility mode is set to >=2.0.0 and we do not
+		// want any 1.x clients to join.
+		"1.0.0": false,
+		"2.0.0": true,
+	},
+	flushMode: {
+		// Note: 1.x clients are compatible with TurnBased flushing, but here we elect to remain on Immediate flush mode
+		// as a work-around for inability to send batches larger than 1Mb. Immediate flushing keeps batches smaller as
+		// fewer messages will be included per flush.
+		"1.0.0": FlushMode.Immediate,
+		"2.0.0": FlushMode.TurnBased,
+	},
+	gcOptions: {
+		"1.0.0": {},
+		// Explicitly disable running Sweep in compat mode "2". Although sweep is supported in 2.x, it is disabled by default.
+		// This setting explicitly disables it to be extra safe.
+		"3.0.0": { enableGCSweep: true },
+	},
+};
+
+/**
+ * Returns the default version-dependent configs for a given compatibility mode.
+ */
+export function getConfigsForCompatMode(
```

Comment:
i'd like to see this function deeply unit tested. you probably want to make versionDependentOptionConfigMap a parameter, it can default to the hard coded map, so we can test any set of defaults against any compatability version.

Reply:
Good call. I have tests in containerRuntime.spec.ts, but they only indirectly call `getConfigsForCompatMode`. I added tests that directly call it in the latest commit and I can add more if needed